### PR TITLE
Request to merge #1609 latch detection to master

### PR DESCRIPTION
**Pull request by margej**
We appreciate your contributing to Verilator.  If this is your first commit, please add your name to docs/CONTRIBUTORS, and read our contributing guidelines in docs/CONTRIBUTING.adoc.

### CODE REVIEW

**Review by wsnyder — COMMENTED** [review]
This is an excellent addition, thanks for working on this.

Please add description of LATCH/NOLATCH to bin/verilator.  Probably a good idea for LATCH to also recommend using always_latch in favor of suppressing the warning, similar to what the warning prints.

File: test_regress/t/t_lint_latch_2.v [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+   input      [1:0] i;
+   output reg [1:0] o;
+
+    // This should not detect a latch as all options are covered
+   always @* begin
+     if      (i==2'b00) o = 2'b11;
+     else if (i==2'b01) o = 2'b10;
+     else if (i==2'b10) o = 2'b01;
+     else if (i==2'b11) o = 2'b00;
```
[the remark this answers is not in the document]
Comment (margej) [reply test_regress/t/t_lint_latch_2.v:16]: I'd like to leave this in. A future update to latch detection could detect the if was "whole" _without_ the else on line 17

File: src/V3Active.cpp [hunk 2]
```diff
@@ -135,6 +291,48 @@ class ActiveNamer final : public ActiveBaseVisitor {
...
+        }
+    }
+    virtual void visit(AstNodeIf* nodep) {
+        LatchDetectGraphVertex* parentp = m_graph.currentp();
+        LatchDetectGraphVertex* branchp = m_graph.addPathVertex(parentp, "BRANCH", true);
+        m_graph.addPathVertex(branchp, "IF");
+        iterateAndNextNull(nodep->ifsp());
+        m_graph.addPathVertex(branchp, "ELSE");
+        iterateAndNextNull(nodep->elsesp());
+        m_graph.currentp(parentp);
```
[the remark this answers is not in the document]
Comment (margej) [reply src/V3Active.cpp:322]: Having previously reviewed a number of .dot graphs for compound if statements when originally working on this, I think it is correct yes.

parentp is saved/restored (unmodified) at entry/exit to each recursive call. addPathVertex advances currentp() to the newly created vertex, so recursive ifs/elses are always attached to the correct parent. As we return back up the call stack, parentp tracks back up towards the root vertex, so when get back to the original else branchp is the same as it was for the original if..... This is difficult to verbalise! maybe i'll attach an example .dot

File: test_regress/t/t_lint_latch_2.v (hunk 1 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply test_regress/t/t_lint_latch_2.v:16]: That's fine.

**Comment by wsnyder** [comment]
I'm ok to merge this, do you think it's ready?

**Comment by margej** [comment]
I think so, but I have not yet added the description of LATCH/NOLATCH to bin/verilator as you suggested

**Comment by wsnyder** [comment]
Ahh, missed that was missing.  Ok, after that once passes I'll merge.

**Comment by margej** [comment]
Good to go I think

**Pull request merged**